### PR TITLE
Secret ownership fixes

### PR DESCRIPTION
**Pull request by PietroPasotti**
- uniform Secret.owner = app | Secret.granted = app
- validate ownership on write paths and grantedness on read paths

### CODE REVIEW

**Comment by PietroPasotti** [comment]
todo: figure out if 'owner=app' means something different in terms of who can read and what non-leaders can read.

**Comment by benhoyt** [comment]
See also https://github.com/canonical/operator/pull/1067. Juju is changing how secret-get works (for secrets owners to work more like secret observers with "refresh" and so on). We're updating ops to that approach (ahead of the Juju change).

**Comment by PietroPasotti** [comment]
@benhoyt added a switch to toggle that behaviour if juju_version<3.2
@tonyandrewmeyer not sure how to generalize this and future-proof for upcoming juju changes. Maybe in time it'd pay off to have separate `_MockModelBackend` classes per juju (major) versions?

**Comment by tonyandrewmeyer** [comment]
> Maybe in time it'd pay off to have separate `_MockModelBackend` classes per juju (major) versions?

The problem I see with that is that it seems like Juju isn't really semver? If it was, then the "owner always peeks" removal would be going into 4.0 and not any 3.x. But if I understand correctly, it's not even being introduced in a minor version but in multiple patch ones. It's before my time, but it looks like `open-port` on k8s was added in a patch not minor, and adding a Pebble version with exec context was added in a patch not minor.

I do think it would be great to have something better than having lots of `if simulated_juju_version` complexity scattered around. Not sure yet what that would be.

**Comment by PietroPasotti** [comment]
> > Maybe in time it'd pay off to have separate `_MockModelBackend` classes per juju (major) versions?
> 
> The problem I see with that is that it seems like Juju isn't really semver? If it was, then the "owner always peeks" removal would be going into 4.0 and not any 3.x. But if I understand correctly, it's not even being introduced in a minor version but in multiple patch ones. It's before my time, but it looks like `open-port` on k8s was added in a patch not minor, and adding a Pebble version with exec context was added in a patch not minor.
> 
> I do think it would be great to have something better than having lots of `if simulated_juju_version` complexity scattered around. Not sure yet what that would be.

Let's keep it in the back of our heads, something will come up eventually

**Comment by PietroPasotti** [comment]
@benhoyt @tonyandrewmeyer I refactored the access control logic based on [my latest findings](https://discourse.charmhub.io/t/secret-access-permissions/12627)

It should be correct now.

**Comment by PietroPasotti** [comment]
One more thing:
this work has revealed that we can simplify considerably the Secret data model: `granted` can go. 
First I realized that there was no difference in view/manage permissions depending on whether the (remote) owner has granted the secret to our app or our unit, since `secret-grant [app]` is equivalent to: `for unit in app.units: secret-grant [unit]`. 

Then I realized we can strip that whole concept completely: it doesn't make sense to put a Secret in `State.secrets` if we're not entitled to know it exists (i.e. if we don't own it nor we've been granted access to it). So the mere presence of a Secret in `State.secrets` tells us we're entitled to `view` it, either because our unit owns it, or our app owns it, or we've been granted it.
If we hadn't been granted it, it wouldn't be part of our State.secrets: the exhaustive list of secrets we are aware of.

This is a BC change so far as the API is concerned, I added a deprecation notice for `Secret.granted`.
However, some tests may break since the idea used to be: put a Secret in State.secrets and make it 'not granted', and now the behaviour will be different.

@tonyandrewmeyer major vbump?

**Comment by tonyandrewmeyer** [comment]
> Then I realized we can strip that whole concept completely: it doesn't make sense to put a Secret in `State.secrets` if we're not entitled to know it exists (i.e. if we don't own it nor we've been granted access to it). So the mere presence of a Secret in `State.secrets` tells us we're entitled to `view` it, either because our unit owns it, or our app owns it, or we've been granted it. If we hadn't been granted it, it wouldn't be part of our State.secrets: the exhaustive list of secrets we are aware of.

:+1: 

> This is a BC change so far as the API is concerned, I added a deprecation notice for `Secret.granted`. However, some tests may break since the idea used to be: put a Secret in State.secrets and make it 'not granted', and now the behaviour will be different.
> 
> @tonyandrewmeyer major vbump?

Imo, yes.

File: scenario/mocking.py [hunk 1]
```diff
@@ -302,6 +303,19 @@ def secret_add(
...
 
+    def _check_can_manage_secret(
+        self,
+        secret: "Secret",
+    ):
+        # FIXME: match real tracebacks
+        if secret.owner is None:
+            raise SecretNotFoundError("this secret is not owned by this unit/app")
+        if secret.owner == "app" and not self.is_leader():
+            raise SecretNotFoundError(
```
Comment (tonyandrewmeyer) [inline scenario/mocking.py:314]: This is the exception when doing `self.model.get_secret(id=id).get_info()` from a unit that's not the leader:

```
unit-testing-secrets-1: 20:04:11 ERROR unit.testing-secrets/1.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/model.py", line 2946, in _run
    result = subprocess.run(args, **kwargs)  # type: ignore
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('/var/lib/juju/tools/unit-testing-secrets-1/secret-get', 'secret://085db680-df30-4f13-82d7-eae3392be3d4/clffid0eq7ms77ta2kbg', '--format=json')' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/./src/charm.py", line 54, in <module>
    ops.main(TestingSecretsCharm)  # type: ignore
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/main.py", line 451, in __call__
    return main(charm_class, use_juju_for_storage=use_juju_for_storage)
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/main.py", line 436, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/main.py", line 144, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/framework.py", line 340, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/framework.py", line 842, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/framework.py", line 931, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/./src/charm.py", line 39, in _on_get_unit_secret_action
    event.set_results({"password": self.model.get_secret(id=id).get_info()})
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/model.py", line 276, in get_secret
    content = self._backend.secret_get(id=id, label=label)
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/model.py", line 3302, in secret_get
    result = self._run('secret-get', *args, return_output=True, use_json=True)
  File "/var/lib/juju/agents/unit-testing-secrets-1/charm/venv/ops/model.py", line 2948, in _run
    raise ModelError(e.stderr) from e
ops.model.ModelError: ERROR permission denied
```

File: scenario/mocking.py [hunk 2]
```diff
@@ -302,6 +303,19 @@ def secret_add(
         self._state.secrets.append(secret)
         return id
 
+    def _check_can_manage_secret(
+        self,
+        secret: "Secret",
+    ):
+        # FIXME: match real tracebacks
+        if secret.owner is None:
+            raise SecretNotFoundError("this secret is not owned by this unit/app")
```
Comment (tonyandrewmeyer) [inline scenario/mocking.py:312]: This is the exception when doing `self.model.get_secret(id="...")` from an application in the same model that somehow magically got hold of an ID of a different app's secret:

```
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/model.py", line 2838, in _run
    result = subprocess.run(args, **kwargs)  # type: ignore
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('/var/lib/juju/tools/unit-testing-actions-0/secret-get', 'secret://xxxx', '--format=json')' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/./src/charm.py", line 102, in <module>
    ops.main(ActionsTestingCharm)  # type: ignore
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/main.py", line 456, in __call__
    return main(charm_class, use_juju_for_storage=use_juju_for_storage)
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/main.py", line 441, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/main.py", line 149, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/framework.py", line 342, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/framework.py", line 839, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/framework.py", line 928, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/./src/charm.py", line 44, in _on_simple_action
    self.model.get_secret(id="secret://xxxx")
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/model.py", line 272, in get_secret
    content = self._backend.secret_get(id=id, label=label)
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/model.py", line 3194, in secret_get
    result = self._run('secret-get', *args, return_output=True, use_json=True)
  File "/var/lib/juju/agents/unit-testing-actions-0/charm/venv/ops/model.py", line 2840, in _run
    raise ModelError(e.stderr)
ops.model.ModelError: ERROR permission denied
```

**Review by tonyandrewmeyer — APPROVED** [review]

File: scenario/mocking.py [hunk 3]
```diff
@@ -311,6 +325,12 @@ def secret_get(
         peek: bool = False,
     ) -> Dict[str, str]:
         secret = self._get_secret(id, label)
+
+        if self._context.juju_version <= "3.2":
+            # in juju<3.2, secret owners always track the latest revision.
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply scenario/mocking.py:330]: makes sense to you @tonyandrewmeyer?

[the remark this answers is not in the document]
Comment (tonyandrewmeyer) [reply scenario/mocking.py]: > AFAIU it is backported to up until 3.1.7, so the check is in fact simpler: all versions below 3.1.7, if they support secrets at all, have the 'old' behaviour. versions above it, the new one.

IIUC:

* 2.x: no secrets - <3.1.7 works
* 3.0.x: has secrets, has 'old' behaviour - <3.1.7 works
* 3.1.0-3.1.6: has secrets, has 'old' behaviour - <3.1.7 works
* 3.1.7-3.1.x: has secrets, has 'new' behaviour - <3.1.7 works
* 3.2.0-3.2.x: has secrets, has 'old' behaviour - *<3.1.7 will not work*
* 3.3.0: has secrets, has 'old' behaviour - *<3.1.7 will not work*
* 3.3.1: has secrets, will have 'new' behaviour - <3.1.7 works
* 3.4.x: has secrets, will have 'new' behaviour - <3.1.7 works

So I think if the goal is to match the Juju behaviour then it needs to be `juju_version < "3.1.7" or juju_version == "3.3.0" or ("3.2.0" <= juju_version < "3.3.0")`

[the remark this answers is not in the document]
Comment (PietroPasotti) [reply scenario/mocking.py]: wait, are you saying they only backported to 3.1.7 but not any other version-with-secrets prior 3.3.1? What/why?

[the remark this answers is not in the document]
Comment (tonyandrewmeyer) [reply scenario/mocking.py]: If I understand correctly, 3.2 is considered closed so isn't getting (non-security?) fixes. 3.3.0 is already out so can't get it, but it'll be in the next 3.3.x, which will be 3.3.1 (pretty soon I think).

[the remark this answers is not in the document]
Comment (PietroPasotti) [reply scenario/mocking.py]: huh. If 3.2 is closed we can also skip supporting it. Same for 3.1 I suppose? So we can make our life easier and say:
3.0 has old behaviour. Anything else has new.

**Pull request merged**